### PR TITLE
fix: support resolveGranularity of ApisixRoute

### PR DESCRIPTION
**Pull request by AlinsRan**
### Type of change:

- [x] Bugfix
- [x] New feature provided
- [x] Improve performance
- [ ] Backport patches

#### ref:
* #1225 

#### fixes: 
* Repair incomplete maintenance of [service->upstream]relation.
* #870
* #1157

### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

**Comment by tao12345666333** [comment]
Also, please adjust your parameter order

````
func Equal(t TestingT, expected, actual interface{}, msgAndArgs ...interface{}) bool
````

File: pkg/apisix/upstreamservicerelation.go [hunk 1]
```diff
@@ -37,79 +39,82 @@ func newUpstreamServiceRelation(c *cluster) *upstreamService {
...
 	}
+	err = u.cluster.cache.DeleteUpstreamServiceRelation(relation)
+	if err != nil {
+		return err
+	}
 	return nil
 }
 
-func (u *upstreamService) Create(ctx context.Context, relation *v1.UpstreamServiceRelation) error {
+// to do: delete one of
```
[the remark this answers is not in the document]
Comment (AlinsRan) [reply pkg/apisix/upstreamservicerelation.go:85]: Delete one of the upstreams. Currently, only service is deleted. There will be some redundant objects, but the results will not be affected.

**Comment by AlinsRan** [comment]
> Also, please adjust your parameter order
> 
> ```
> func Equal(t TestingT, expected, actual interface{}, msgAndArgs ...interface{}) bool
> ```

Done.

File: pkg/types/apisix/v1/types.go [hunk 2]
```diff
@@ -468,17 +470,20 @@ func NewDefaultPluginConfig() *PluginConfig {
...
 	var p []byte
-	if subset == "" {
-		p = make([]byte, 0, len(namespace)+len(name)+len(pstr)+2)
-	} else {
-		p = make([]byte, 0, len(namespace)+len(name)+len(subset)+len(pstr)+3)
+	plen := len(namespace) + len(name) + len(pstr) + 2
+	if subset != "" {
+		plen = plen + len(subset) + 1
+	}
+	if resolveGranularity != "" && resolveGranularity != types.ResolveGranularity.Endpoint {
```
Comment (tao12345666333) [inline pkg/types/apisix/v1/types.go:482]:
```suggestion
	if resolveGranularity != types.ResolveGranularity.Service {
```

Should we change this?

File: pkg/types/apisix/v1/types.go [hunk 3]
```diff
@@ -489,6 +494,10 @@ func ComposeUpstreamName(namespace, name, subset string, port int32) string {
 		buf.WriteByte('_')
 	}
 	buf.WriteString(pstr)
+	if resolveGranularity != "" && resolveGranularity != types.ResolveGranularity.Endpoint {
```
Comment (tao12345666333) [inline pkg/types/apisix/v1/types.go:497]: Ditto

**Comment by tao12345666333** [comment]
I want you can add an e2e test case to cover #870, thanks!

File: pkg/types/apisix/v1/types.go (hunk 2 above)
Comment (AlinsRan) [reply pkg/types/apisix/v1/types.go:482]: I think it should be `resolveGranularity == types.ResolveGranularity.Service`, but should we support more ResolveGranularity?

**Review by tao12345666333 — APPROVED** [review]
LGTM

thanks

File: pkg/apisix/resource_test.go [hunk 4]
```diff
@@ -86,8 +86,11 @@ func TestItemConvertUpstream(t *testing.T) {
 	ups, err := ite.upstream()
 	assert.Nil(t, err)
 	assert.Len(t, ups.Nodes, 2)
-	assert.Equal(t, ups.Nodes[0], v1.UpstreamNode{Host: "httpbin.org", Port: 80, Weight: 1})
-	assert.Equal(t, ups.Nodes[1], v1.UpstreamNode{Host: "foo.com", Port: 8080, Weight: 2})
+	if ups.Nodes[0].Host == "foo.com" {
+		ups.Nodes[0], ups.Nodes[1] = ups.Nodes[1], ups.Nodes[0]
```
Comment (lingsamuel) [inline pkg/apisix/resource_test.go:90]: what's this?

File: pkg/providers/apisix/apisix_upstream.go [hunk 5]
```diff
@@ -266,55 +266,66 @@ func (c *apisixUpstreamController) sync(ctx context.Context, ev *types.Event) er
 		clusterName := c.Config.APISIX.DefaultClusterName
 		for _, port := range svc.Spec.Ports {
 			for _, subset := range subsets {
-				upsName := apisixv1.ComposeUpstreamName(namespace, name, subset.Name, port.Port)
 				// TODO: multiple cluster
-				ups, err := c.APISIX.Cluster(clusterName).Upstream().Get(ctx, upsName)
-				if err != nil {
-					if err == apisixcache.ErrNotFound {
-						continue
+				update := func(upsName string) error {
```
Comment (lingsamuel) [inline pkg/providers/apisix/apisix_upstream.go:270]: Can it be a function instead of a variable?

File: pkg/providers/k8s/endpoint/base.go [hunk 6]
```diff
@@ -131,3 +131,23 @@ func (c *baseEndpointController) syncEndpoint(ctx context.Context, ep kube.Endpo
...
 	return nil
 }
+
+func (c *baseEndpointController) syncEmptyEndpoint(ctx context.Context, ep kube.Endpoint) error {
+	namespace, err := ep.Namespace()
+	if err != nil {
+		return err
+	}
+	svcName := ep.ServiceName()
+	log.Infow("The syncEndpoint, service has been deleted, try to update upstream",
```
Comment (lingsamuel) [inline pkg/providers/k8s/endpoint/base.go:141]:
```suggestion
	log.Debugw("The service has been deleted, try to delete upstream relation",
```
Potential frequently called function trace logging should be debug level.

File: pkg/providers/k8s/endpoint/base.go [hunk 7]
```diff
@@ -131,3 +131,23 @@ func (c *baseEndpointController) syncEndpoint(ctx context.Context, ep kube.Endpo
...
+		zap.String("Namespace", namespace),
```
Comment (lingsamuel) [inline pkg/providers/k8s/endpoint/base.go:142]: We use `snake_case` style in log.

File: pkg/apisix/resource_test.go (hunk 4 above)
Comment (AlinsRan) [reply pkg/apisix/resource_test.go:90]: The position values of 1 and 0 may be different.

File: pkg/providers/k8s/endpoint/base.go [hunk 8]
```diff
@@ -131,3 +131,23 @@ func (c *baseEndpointController) syncEndpoint(ctx context.Context, ep kube.Endpo
...
+		zap.String("ServiceName", svcName),
+	)
+	clusterName := c.Config.APISIX.DefaultClusterName
+	err = c.APISIX.Cluster(clusterName).UpstreamServiceRelation().Delete(ctx, namespace+"_"+svcName)
+	if err != nil {
+		log.Errorw("try to delete upstream relation and upstream nodes failed",
```
Comment (lingsamuel) [inline pkg/providers/k8s/endpoint/base.go:148]:
```suggestion
		log.Errorw("delete upstream relation failed",
```

While this function does delete nodes, the log doesn't need to display it.

File: pkg/types/apisix/v1/types.go [hunk 9]
```diff
@@ -468,17 +470,20 @@ func NewDefaultPluginConfig() *PluginConfig {
...
-	if subset == "" {
-		p = make([]byte, 0, len(namespace)+len(name)+len(pstr)+2)
-	} else {
-		p = make([]byte, 0, len(namespace)+len(name)+len(subset)+len(pstr)+3)
+	plen := len(namespace) + len(name) + len(pstr) + 2
+	if subset != "" {
+		plen = plen + len(subset) + 1
+	}
+	if resolveGranularity == types.ResolveGranularity.Service {
+		plen = plen + len(resolveGranularity) + 1
```
Comment (lingsamuel) [inline pkg/types/apisix/v1/types.go:483]: The function comment should also be updated to show that resolveGranularity is not composited in the upstream name when it is endpoint.

File: pkg/apisix/upstreamservicerelation.go [hunk 10]
```diff
@@ -37,79 +39,79 @@ func newUpstreamServiceRelation(c *cluster) *upstreamService {
...
-		if err != nil {
-			return err
+			continue
 		}
 	}
 	return nil
 }
 
-func (u *upstreamService) Create(ctx context.Context, relation *v1.UpstreamServiceRelation) error {
+// to do: delete one of
```
[the remark this answers is not in the document]
Comment (AlinsRan) [reply pkg/apisix/upstreamservicerelation.go:82]: I want to use svc delete to replace it, and I don't want it to invade more code.

[the remark this answers is not in the document]
Comment (lingsamuel) [reply pkg/apisix/upstreamservicerelation.go:82]: Why not write this in TODO to make it complete?

File: pkg/providers/apisix/apisix_upstream.go (hunk 5 above)
Comment (AlinsRan) [reply pkg/providers/apisix/apisix_upstream.go:270]: Yes, but it has many contexts, requires more parameters, and can't be reused. I do not recommend this.

File: pkg/types/apisix/v1/types.go [hunk 11]
```diff
@@ -468,17 +470,21 @@ func NewDefaultPluginConfig() *PluginConfig {
...
 	var p []byte
-	if subset == "" {
-		p = make([]byte, 0, len(namespace)+len(name)+len(pstr)+2)
-	} else {
-		p = make([]byte, 0, len(namespace)+len(name)+len(subset)+len(pstr)+3)
+	plen := len(namespace) + len(name) + len(pstr) + 2
+	if subset != "" {
+		plen = plen + len(subset) + 1
+	}
+	// the resolveGranularity is not composited in the upstream name when it is endpoint.
```
Comment (lingsamuel) [inline pkg/types/apisix/v1/types.go:482]: This comment describes the behavior of the function and should be placed in the comment of the function signature, not in the function body.

Comment (AlinsRan) [reply pkg/types/apisix/v1/types.go]: My negligent. Resolved.

**Review by lingsamuel — APPROVED** [review]

**Pull request merged**